json-schema-generator/src/main/java/io/micronaut/jsonschema/generator/SourceGenerator.java

```diff
@@ -0,0 +1,638 @@
+            inputFileName = getInputFileName() != null ? getInputFileName() : config.getInputName();
+            // single java object is generated
+            if (config.outputFileName() != null && !config.outputFileName().isBlank()) {
+                var outputFileName = config.outputFileName();
+                // remove extension from file name if there is
+                if (config.outputFileName().contains(".")) {
+                    outputFileName = outputFileName.substring(0, outputFileName.indexOf('.'));
+                }
+                return generateFromSchema(jsonSchema, config.outputPath(), config.outputPackageName(), outputFileName);
+            } else {
+                saveDefinitions(jsonSchema);
+                return generateDefinitions(jsonSchema, config.outputPath(), config.outputPackageName());
+            }
+        }
+        return null;
+    }
+
+    /**
+     * A method for creating multiple objects (class, record, interface) from a folder of JSON Schema.
+     *
+     * @param config     The SourceGeneratorConfig
+     */
+    private void generateFolder(SourceGeneratorConfig config) throws IOException {
+        HashMap<Schema, String> schemas = new HashMap<>();
+        Path jsonFolder =  config.inputFolder();
+        // Walk through the directory to find all json files
+        try (Stream<Path> paths = Files.walk(jsonFolder).filter(file -> file.toString().endsWith(".schema.json"))) {
+            paths.forEach(path -> {
+                // Read content of each JSON file
+                var jsonSchema = new FileLoader(path.toFile()).load();
+                assert jsonSchema != null;
+                inputFileName = path.toString().substring(jsonFolder.toString().length() + 1);
+                schemas.put(jsonSchema, inputFileName);
+                saveDefinitions(jsonSchema);
+            });
+        } catch (IOException e) {
+            throw new FileSystemException(jsonFolder.toString());
+        }
+
+        schemas.forEach((jsonSchema, fileName) -> {
+            try {
+                inputFileName = fileName;
+                generateDefinitions(jsonSchema, config.outputPath(), config.outputPackageName());
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+
+    private void saveDefinitions(Schema jsonSchema) {
+        String schemaName = jsonSchema.hasTitle() ? jsonSchema.getTitle() : inputFileName.substring(0, inputFileName.indexOf('.'));
+        String finalSchemaName = getClassName(schemaName);
+
+        // save all definition and oneOf types
+        if (jsonSchema.hasOneOf()) {
+            jsonSchema.getOneOf().forEach(oneOf -> {
+                if (oneOf.has$ref()) {
+                    String ref = oneOf.get$ref();
+                    if (ref.indexOf("#") == 0) {
+                        ref = inputFileName + ref;
+                    }
+                    context.addOneOf(ref);
```

Comment:
Why can't we generate these classes here?

Reply:
Because definitions might include properties that point to other references we have not seen yet.